Refactor dedicated server management script and docker container

turtletowerz opened the pull request:
closes #2650 

There were quite a few areas where the previous versions of the script and container fell short, notably:
* Symlinks hacks to get the Docker container to work with the old system of symlinking
* The base steamcmd/steamcmd docker image had steamcmd installed twice, adding ~30% to the image size
* There was no easy way to run commands or attach to the Docker Container
* Simarily, an entrypoint script to was added catch SIGTERM and gracefully shutdown to prevent previous data loss issues
* The regular management script (not Docker) had problems with the backup system
* Documentation was not very clear and led to some confusion; Various Discord forum posts were already answered in the docs. 

There was also some smaller fixes and general code cleanup to increase the overall quality.

Once this is merged the [wiki page for servers](https://github.com/tModLoader/tModLoader/wiki/Starting-a-modded-server) will be updated with the new information

turtletowerz committed:
fix dotnet force reinstalling on server start

turtletowerz committed:
Merge branch '1.4.4' of https://github.com/turtletowerz/tModLoader into 1.4.4

turtletowerz committed:
refactor managescript script and docker container

* utilize docker bind mounts more effectively and remove symlink hacks

* remove dependency on steamcmd/steamcmd docker image to reduce image size by 30%

* add tmux support to allow for commands to be run from `docker exec`

* add entrypoint script to catch SIGTERM and gracefully shutdown

* fix non-Docker backups to actually work

* update documentation to reflect above changes

turtletowerz committed:
add information on how to run commands in the docker container

scmarlatt commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/docker-compose.yml:
```diff
@@ -3,17 +3,19 @@ version: "3.0"
       - 7777:7777
     volumes:
-      - ./tModLoader:/home/tml/.local/share/Terraria/tModLoader
+      - ./tModLoader:/data
```
I came across an issue on macOS Sierra 15.1 and had to specify `platform: linux/amd64`

turtletowerz replied:
Added platform as an option, but noted that there ARM is not functional currently

scmarlatt commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/docker-compose.yml:
```diff
@@ -3,17 +3,19 @@ version: "3.0"
-    stdin_open: true
+      # args:
+      #   TMLVERSION: v2023.8.3.3
+      #   UID: 1000
+      #   GID: 1000
```
Again, might be macOS specific, but when trying to build the container, it said the values I obtained from running `id` were already in use.

Just passing 1000/1000 seemed to allow the build to progress though

scmarlatt replied:
Again, might be macOS specific, but when trying to build the container, it said the values I obtained from running id were already in use.

Just passing 1000/1000 seemed to allow the build to progress though

turtletowerz committed:
Merge branch 'tModLoader:1.4.4' into 1.4.4

turtletowerz committed:
update README and fix workshop error

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/README.md:
```diff
@@ -1,111 +1,121 @@
 If you want to run tModLoader without needing any input on startup (such as from an init system), copy the example [serverconfig.txt](https://github.com/tModLoader/tModLoader/tree/1.4.4/patches/tModLoader/Terraria/release_extras/serverconfig.txt) and change the settings how you like. Key options are defined below, and other options can be found [on the Terraria wiki](https://terraria.wiki.gg/wiki ...
 * `worldname` changes the default world name when creating a new world using autocreate **You do not need to include .wld in your world name**. This setting **will not** work with an existing world, see the `world` option for an existing world
-* `world` sets the exact path to an existing or new terraria world, ex. `worldpath/to/your/world.wld`. **For Docker installations**, the world path must follow `/home/tml/.local/share/Terraria/tModLoader/Worlds/your_world.wld`
+* `world` sets the exact path to an existing or new terraria world, ex. `worldpath/to/your/world.wld`. **For Docker installations**, the world path must follow `/data/Worlds/your_world.wld`
```
Hmm, the fact that it's randomly `/data` inside the container is a bit painful. It'd be nice if paths inside weren't so unguessable/invisible. Can it be `tModLoader/Worlds/your_world.wld`?

turtletowerz replied:
Agreed, I've committed a small change to fix `/data` to `/tModLoader`

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/Dockerfile:
```diff
@@ -1,44 +1,126 @@
+
+# Use wait instead of `tmux wait tml &> /dev/null` so that trapping the shutdown works properly
+wait \${!}
+EOF
```
Why do these scripts need to be part of the container? Can't people use tmux outside the docker run command? This creates a layer of complexity and separation between users of the management script and docker users

turtletowerz replied:
tmux makes it significantly easier for graceful Docker shutdowns. I had some issues in the past with tModLoader not shutting down gracefully in Docker and this was the simplest solution. I can do more testing to see if there's a way to solve this.

Additionally adding tmux allows for users to do one-off commands without having to enter the container.

For example: `docker exec -it tml execute "say Hello World!"` instead of having to `docker attach tml` and then type `say Hello World!" in the console.

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -69,114 +59,92 @@ function update_script {
 	echo "Updating from version v$script_version to v$latest_script_version"
-	curl -s -O "$script_url" 2>/dev/null || wget -q "$script_url"
-	mv manage-tModLoaderServer.sh.1 manage-tModLoaderServer.sh
+	mv manage-tModLoaderServer.sh manage-tModLoaderServer.sh.old
+	curl -s -O "$script_url" || exit 1
```
Why change this?

turtletowerz replied:
A misunderstanding on how `curl` deals with duplicate names. I've reverted my commit for this line, but I still think it's better to keep the curl output with `|| exit 1` as we should not ignore if there's an issue when downloading management script updates

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -185,79 +153,85 @@ function install_tml_github {
-	# Make folder structure
 	if ! is_in_docker; then
 		echo "Creating folder structure"
-		mkdir -p Mods Worlds logs
```
Where go logs?

turtletowerz replied:
For non-Docker setups the logs will be in `server/tModLoader-Logs` as it is with any normal installation. Do you think I should symlink them so that `./logs -> ./server/tModLoader-Logs`?

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -279,34 +253,33 @@ function print_help {
- --steamcmdpath      Custom steamcmd path for users who do not have a installation in PATH.
- --keepbackups       Will keep all tML backups instead of the most recent one when updating
+ --keepbackups       When installing with --github, keep all previous versions instead of deleting them when updating
+ --tmlversion        Version of tModLoader to install. Only works if --github is provided. Functionally equivalent to the TMLVERSION env variable
```
Should probably keep the legacy arg as well, no need to break things unnecessarily

turtletowerz replied:
Brought back `--tml-version` to keep compatibility. All the other flags are the same I just changed the flag descriptions

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -279,34 +253,33 @@ function print_help {
 start_args=""
 
-# Check for updates to the script if it's not running in a Docker container
-update_script
```
no auto update checker?

turtletowerz replied:
It was moved below flag parsing so that doing something like `-h` or `-v` would not prompt for an update since there's no need in that case.

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -324,101 +297,86 @@ while [[ $# -gt 0 ]]; do
-				steamcmd=false
-				shift
-			fi
+		--tmlversion)
```
Why this not set `--github` anymore? breaking change?

turtletowerz replied:
Must've accidentally missed that, I have brought it back

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -324,101 +297,86 @@ while [[ $# -gt 0 ]]; do
-		try_make_link "$folder/banlist.txt" "$folder/server/banlist.txt"
-
-		# Provide option to use custom argsConfig file
-		if [[ -f "$folder/cli-argsConfig.txt" ]]; then
-			ln -s "$folder/cli-argsConfig.txt" "$folder/server/cli-argsConfig.txt"
-		fi
```
Where did this go?

turtletowerz replied:
For docker it was moved to be a bind mount, so in `docker-compose.yml`

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -41,10 +41,10 @@ echo "Verifying .NET...."  2>&1 | tee -a "$LogFile"
 
 # Get Dotnet Version expecting to have installed
-source ./DotNetVersion.sh
+. ./DotNetVersion.sh
```
Why?

turtletowerz replied:
`source` and `.` are functionally equivalent in bash and `.` is used elsewhere in the same script so I changed it to match

Chicken-Bones replied:
Lets revert changes to this file, I've suggested a log solution above

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -324,101 +297,86 @@ while [[ $# -gt 0 ]]; do
 	start)
-		if ! [[ -f "$folder/server/start-tModLoaderServer.sh" ]]; then
+		# Edge-case for ScriptCaller.sh where dotnet exists but TML logs don't yet
+		export SKIP_DOTNET_LOGCHECK=1
```
shouldn't this be docker only?

turtletowerz replied:
No, this is an issue with any install from Github. One solution I had as a temporary fix was to create `server.log` and `client.log` manually during the `install-tml` phase but I don't really like the idea of creating log files before anything is actually logged. If you think I should do this instead of `SKIP_DOTNET_LOGCHECK` then I can change that

Chicken-Bones replied:
Yeah, lets generate a `server.log` of `dotnet installed via management script... pending first server start...`

Chicken-Bones reviewed (changes requested):
Concerned about breaking changes. Needs user testing.

turtletowerz committed:
Merge branch 'tModLoader:1.4.4' into 1.4.4

turtletowerz committed:
Use tModLoader instead of data for data path name

turtletowerz committed:
revert script update changes

turtletowerz committed:
keep --tml-version flag for compatibility

turtletowerz committed:
fix issue where setting --tmlversion would not automatically enable github

turtletowerz committed:
Merge branch '1.4.4' of https://github.com/turtletowerz/tModLoader into 1.4.4

Tredecate committed:
Make `cd` reference the current script when ran with `source`, rather than parent script

Tredecate committed:
Make manage script call `ScriptCaller.sh` directly via `source`

This is one step towards preserve the PID of the management script and allow for docker attach and signal propagation to work correctly.

Tredecate committed:
Small formatting fix

Tredecate committed:
Remove tmux from dockerfile
- Removes the need for tmux and related functions
- Sets entrypoint to use updated management script
- Adds required flags to docker-compose to enable attaching

turtletowerz committed:
Merge pull request #1 from Tredecate/docker-script-rework

Management Script Adjustments

turtletowerz committed:
Fix downloading to folder with spaces, unify start implementation and small fixes

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/README.md:
```diff
@@ -98,12 +99,14 @@ To explore all the options before continuing, run `./manage-tModLoaderServer.sh
-3. Install any necessary mods
-   * Mods will be automatically installed during the tModLoader installation step, but can also be installed separately by running `./manage-tModLoaderServer.sh install-mods`. No mods will be installed if `install.txt` is missing, and no mods will be enabled if `enabled.json` is missing. **You will need a `Mods/enabled.json` to contain all Mods that you want enabled, including local mods**. 
+3. Install any necessary mods with `./manage-tModLoaderServer.sh install-mods`
+   * Run . No mods will be installed if `install.txt` is missing, and no mods will be enabled if `enabled.json` is missing. **You will need a `Mods/enabled.json` to contain all Mods that you want enabled, including local mods**
```
Run?

turtletowerz committed:
Revert global ScriptCaller change and move it to management script

turtletowerz committed:
attempt fix for ScriptCaller forcing dotnet reinstall

Chicken-Bones reviewed (approved):
Approved pending user testing

turtletowerz committed:
Update README.md

turtletowerz committed:
Merge branch '1.4.4' into 1.4.4

oksoff commented on patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/manage-tModLoaderServer.sh:
```diff
@@ -324,101 +299,89 @@ while [[ $# -gt 0 ]]; do
-			ln -s "$folder/cli-argsConfig.txt" "$folder/server/cli-argsConfig.txt"
-		fi
+		if is_in_docker; then
+			mkdir -p "$folder/Mods" "$folder/Worlds"
```
This should've been
```suggestion
			mkdir -p "Mods" "Worlds"
```
Since we're already in `$folder` from :347

It's only a problem if `folder` is defined though.

EDIT: or rather it's a problem if `folder` is not an absolute path.
